src/relentless/simulate/dilute.py

```diff
@@ -171,43 +171,54 @@ def process(self, sim, sim_op):
-                gr = ens.rdf[a, b].table[:, 1]
-
-                # only count continuous range of finite values
-                flags = numpy.isinf(u)
-                first_finite = 0
-                while flags[first_finite] and first_finite < len(r):
-                    first_finite += 1
-                r = r[first_finite:]
-                u = u[first_finite:]
-                f = f[first_finite:]
-                gr = gr[first_finite:]
-
-                if sim.dimension == 3:
-                    geo_prefactor = 4 * numpy.pi * r**2
-                elif sim.dimension == 2:
-                    geo_prefactor = 2 * numpy.pi * r
-                else:
-                    raise ValueError(
-                        "Geometric integration factor unknown for extent type"
-                    )
-                y = (geo_prefactor / 6.0) * rho_a * rho_b * f * gr * r
-                ens.P += numpy.trapz(y, x=r)
-
-        sim[self]["ensemble"] = ens
-        sim[self]["num_thermo_samples"] = None
-        sim[self]["num_rdf_samples"] = None
+        B = 0.0
+        N = sum(ens.N[i] for i in sim.types)
+        for i, j in sim.pairs:
+            counting_factor = 2 if i != j else 1
+            r = sim.potentials.pair.linear_space
+            u = sim.potentials.pair.energy((i, j))
+
+            # only count continuous range of finite values
+            flags = numpy.isinf(u)
+            first_finite = 0
+            while flags[first_finite] and first_finite < len(r):
+                first_finite += 1
+            r = r[first_finite:]
+            u = u[first_finite:]
+
+            if sim.dimension == 3:
+                geo_prefactor = 4 * numpy.pi * r**2
+            elif sim.dimension == 2:
+                geo_prefactor = 2 * numpy.pi * r
+            else:
+                raise ValueError("Geometric integration factor unknown for extent type")
+            y_i = ens.N[i] / N
+            y_j = ens.N[j] / N
+            b_ij = -counting_factor * (geo_prefactor / 2.0) * (numpy.exp(-u / kT) - 1)
+            B += y_i * y_j * numpy.trapz(b_ij, x=r)
+        # calculate pressure if no barostat
+        if ens.P is None:
+            rho = N / ens.V.extent
+            ens.P = kT * (rho + B * rho**2)
+            sim[self]["ensemble"] = ens
+            sim[self]["num_thermo_samples"] = None
+            sim[self]["num_rdf_samples"] = None
+        # adjust extent to maintain pressure if barostat
+        else:
+            coeffs = numpy.array([-ens.P / kT, 1, B])
+            v = numpy.max(numpy.roots(coeffs))
+            V = v * N
+            L = V ** (1 / sim.dimension)
+
+            if sim.dimension == 3:
+                ens.V = extent.Cube(L)
+            elif sim.dimension == 2:
+                ens.V = extent.Square(L)
+
+            sim[self]["ensemble"] = ens
+            sim[self]["num_thermo_samples"] = None
+            sim[self]["num_rdf_samples"] = None
 
 
 class Dilute(simulate.Simulation):
```

Comment:
We should update the documentation of this class while we are here:

- [ ] After the equation of g(r), add a paragraph about how the virial equation of state is used. Give the expression for the pressure and the second virial coefficient.
- [ ] Remove the last paragraph. It is out of data, and an implementation detail that is not important anyway.